### PR TITLE
GRW-1421 / feat / Warn if date is too soon

### PR DESCRIPTION
**Pull request by simonauner**
## What?

1. Change the label of the date selector when Switcher **and** Car to "Select date" instead of "When current insurance expires"
2. Warn the user if the selected date is "too soon" (less than 7 days into the future). 

## Why?

This was originally going to be a Slack post but I guess things will be better documented for future reference here.

**Background:**

We cannot guarantee that a Car insurance with start date less than one week in the future can be switched by BankSignering on time. We want to prompt users if the date is too soon. 

**Current state:**

We have the date picker for when an insurance should start on the Offer page. When creating a car quote, the start date is returned as `null` from BE, which by default is turned into `new Date()` in the Offer page. This is not the case for switchers. Prior to this PR (and after, still), `null` dates for a switcher indicates "when my old insurance expires", and `null` is an allowed value for switchers.

**Problem:**

Since we don't know the expiration date of the old insurance we cannot guarantee that we can switch the user in time, thus we cannot allow the users to select "When my old insurance expires". The user _must_ guess the date so that we can prompt them if the date is too soon. 

We also can't show the warning by default, that would be an annoying experience, as it is designed for this PR. (Maybe we should reconsider the design!)

**Solution:**

1. If Car and Switcher, show a different form label that says "Select date" instead of "When old insurance expires". We need to once again make components aware of that they're rendering something for Car.
2. Disallow `null` for Car switchers

**Other thoughts:**

* We cannot guarantee a switch for car within 7 days - but we can for other insurances 🤔 ? This logic should probably apply to all switcher cases?
* What happens when we implement Insurely for Car? Then we will get the same data as we have for switchers for other insurances? Maybe much of this PR will change then anyway?

⚠️ I have not implemented a validation that says that `null` is not allowed for Car switchers.

## Pics

The label:
image

We cannot allow `null` for Car switchers (but for other types of switchers):
image

Warning notice:
image

**Ticket(s): [GRW-1421]**

[GRW-1421]: https://hedvig.atlassian.net/browse/GRW-1421?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/client/components/ModalNew.tsx [hunk 1]
```diff
@@ -95,6 +95,18 @@ export const Modal = ({
 
   const hasOnCloseFunction = Boolean(onClose)
 
+  useEffect(() => {
+    const handleKeyDown = (e: KeyboardEvent) => {
+      if (e.key === 'Escape' && !!onClose) {
```
Comment (alebedev) [inline src/client/components/ModalNew.tsx:100]: Nit: can use optional chaining instead
```
if (e.key === 'Escape') {
  onClose?.()
}
```

File: src/client/components/ModalNew.tsx [hunk 2]
```diff
@@ -95,6 +95,18 @@ export const Modal = ({
...
+        onClose()
+      }
+    }
+    window.addEventListener('keydown', handleKeyDown)
+    return () => {
+      window.removeEventListener('keydown', handleKeyDown)
+    }
+  }, [hasOnCloseFunction, onClose])
```
Comment (alebedev) [inline src/client/components/ModalNew.tsx:108]: hasOnCloseFunction isn't used in this hook.  Let's investigate why this isn't a lint error, it should be

File: src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx [hunk 3]
```diff
@@ -158,13 +173,13 @@ const DateInputModalWrapper = styled.div<{
...
   startDate: string | null,
-  currentInsurer?: CurrentInsurer,
+  isSwitcher?: boolean,
 ) => {
   if (startDate) {
     return parse(startDate, 'yyyy-MM-dd', new Date())
   }
 
-  if (currentInsurer?.switchable) {
+  if (isSwitcher) {
```
Comment (alebedev) [inline src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:182]: Let's comment why

File: src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx [hunk 4]
```diff
@@ -207,16 +222,20 @@ const DateForm = ({
...
+    getDefaultDateValue(startDate, isSwitcher),
   )
   const [dateLocale, setDateLocale] = useState<Locale | null>(null)
 
   const [datePickerOpen, setDatePickerOpen] = useState(false)
 
   useEffect(() => {
-    setDateValue(getDefaultDateValue(startDate, currentInsurer))
-  }, [startDate, currentInsurer])
+    setDateValue(getDefaultDateValue(startDate, isSwitcher))
```
Comment (alebedev) [inline src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:237]: Nit: dateValue is calculated from props, we never set it other way.  We can probably simplify code by calculating it and removing the state

Comment (alebedev) [inline src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:237]: Nit: this file is already huh

File: src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx [hunk 5]
```diff
@@ -424,6 +462,31 @@ export const StartDate = ({
   )
 }
 
+type TooSoonSwitcherErrorModalProps = {
```
Comment (alebedev) [inline src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:465]: Nit: this file is already huge, maybe we can split most of modal markup and styles into separate module?

**Review by alebedev — COMMENTED** [review]
Looks good, some potential future improvements in comments

File: src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx (hunk 5 above)
Comment (alebedev) [inline src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:465]: Nit: this file is already huge, maybe we can split error modal and related utils to separate module?

File: src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx (hunk 4 above)
Comment (simonauner) [reply src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:237]: Nice, didn't even notice 👍  Will fix!

File: src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx (hunk 5 above)
Comment (simonauner) [reply src/client/pages/Offer/Introduction/Sidebar/StartDate.tsx:465]: Sure, that's non unreasonable 🙃

File: src/client/components/ModalNew.tsx (hunk 1 above)
Comment (simonauner) [reply src/client/components/ModalNew.tsx]: Ah, yes, forgot about that one, nice!

File: src/client/components/ModalNew.tsx (hunk 2 above)
Comment (simonauner) [reply src/client/components/ModalNew.tsx]: Yeah, didn't even think about it!

**Review by alebedev — APPROVED** [review]
Looks good!

**Pull request merged**